### DISCUSSION TITLE
vd6.l.qq.com v.gdt.qq.com等域名返回结果为NXDOMAIN

### DISCUSSION THREAD

**Discussion by yyysuo** [IrineSistiana/mosdns#540] [Q&A] [answered]
求各位大神指点一下问题排查的方向。

![全板dns流程+ MOSDNS](https://user-images.githubusercontent.com/10081071/198211847-ca1f68a8-8046-4620-9ba5-326eabe71aa7.jpg)

dns套娃架构如上，主路由IP是10.10.10.1，使用方案2过mosdns时，下列域名返回结果为NXDOMAIN，结果就是腾讯NBA的视频播放缓慢，甚至无法播放；使用方案1走dnsmasq时，返回的IP是正常的，播放也没问题。在ssrp和passwall下均能重现。

出现问题的域名发现的主要是qq相关的
vd6.l.qq.com
i.match.qq.com
l.qq.com
v.gdt.qq.com
i.gdt.qq.com
pingfore.qq.com
jsqmt.qq.com
trace.qq.com
aegis.qq.com

mosdns配置如下：
log:
  level: info
  file: "/tmp/mosdns.log"

include: []

data_providers:
  - tag: geoip
    #file: "/etc/mosdns/geoip.dat"        #浣跨敤cn.dat 楂樻€ц兘
    file: "/usr/share/v2ray/geoip.dat"        #浣跨敤cn.dat 楂樻€ц兘
    auto_reload: true

  - tag: geosite
    #file: "/etc/mosdns/geosite.dat"
    file: "/usr/share/v2ray/geosite.dat"
    auto_reload: true

  - tag: whitelist
    file: "/etc/mosdns/rule/whitelist.txt"
    auto_reload: true

  - tag: blocklist
    file: "/etc/mosdns/rule/blocklist.txt"
    auto_reload: true

  - tag: hosts
    file: "/etc/mosdns/rule/hosts.txt"
    auto_reload: true

  - tag: redirect
    file: "/etc/mosdns/rule/redirect.txt"
    auto_reload: true

plugins:
  - tag: lazy_cache
    type: cache
    args:
      size: 200000
      lazy_cache_ttl: 259200

  - tag: modify_ttl
    type: ttl
    args:
      minimal_ttl: 0
      maximum_ttl: 0

  - tag: "forward_local"
    type: fast_forward
    args:
      upstream:
        - addr: 10.10.10.1:8055

  - tag: "forward_remote"
    type: fast_forward
    args:
      upstream:
        - addr:  10.10.10.1:5335
        #- addr: tls://1.1.1.1
        #- addr: "https://dns.google/dns-query"
        #  dial_addr: "8.8.8.8:443"
        #  idle_timeout: 30
        #  trusted: true
        #- addr: "https://cloudflare-dns.com/dns-query"
        #  dial_addr: "1.1.1.1:443"
        #  idle_timeout: 30
        #  trusted: true

  - tag: query_is_whitelist_domain
    type: query_matcher
    args:
      domain:
        - "provider:whitelist"

  - tag: query_is_blocklist_domain
    type: query_matcher
    args:
      domain:
        - "provider:blocklist"

  - tag: query_is_hosts_domain
    type: hosts
    args:
      hosts:
        - "provider:hosts"

  - tag: query_is_redirect_domain
    type: redirect
    args:
      rule:
        - "provider:redirect"

  - tag: query_is_local_domain
    type: query_matcher
    args:
      domain:
        - "provider:geosite:cn"

  - tag: query_is_non_local_domain
    type: query_matcher
    args:
      domain:
        - "provider:geosite:geolocation-!cn"

  - tag: response_has_local_ip
    type: response_matcher
    args:
      ip:
        - "provider:geoip:cn"

  - tag: query_is_ad_domain
    type: query_matcher
    args:
      domain:
        - "provider:geosite:category-ads-all"

  - tag: match_qtype65
    type: query_matcher
    args:
      qtype: [65]

  - tag: "main_sequence"
    type: "sequence"
    args:
      exec:
        - query_is_hosts_domain
        - query_is_redirect_domain

        - if: query_is_whitelist_domain
          exec:
            - forward_local
            - modify_ttl
            - _return

        - if: "query_is_blocklist_domain || query_is_ad_domain || match_qtype65"
          exec:
            - _new_nxdomain_response
            - _return

        - lazy_cache

        - if: query_is_local_domain
          exec:
            - forward_local
            - modify_ttl
            - _return

        - if: query_is_non_local_domain
          exec:
            - _prefer_ipv4
            - forward_remote
            - modify_ttl
            - _return
        - primary:
            - forward_local
            - if: "(! response_has_local_ip) && [_response_valid_answer]"
              exec:
                - _drop_response
          secondary:
            - _prefer_ipv4
            - forward_remote
            - modify_ttl
          fast_fallback: 200

servers:
  - exec: main_sequence
    listeners:
      - protocol: udp
        addr: ":5454"
      - protocol: tcp
        addr: ":5454"

**Comment by BozhanL** [comment, marked as answer]
把这段删了试试
```yaml
- if: "query_is_blocklist_domain || query_is_ad_domain || match_qtype65"
  exec:
    - _new_nxdomain_response
    - _return
```

**Comment by yyysuo** [comment]
> aegis.qq.com

感谢，问题解决！geosite里面有去广告规则么？这些dns请求也不是type65类型的呀。

**Reply by BozhanL** [reply]
检查一下```query_is_blocklist_domain```和```query_is_ad_domain```的内容

**Reply by yyysuo** [reply]
> 检查一下`query_is_blocklist_domain`和`query_is_ad_domain`的内容

  - tag: query_is_ad_domain
    type: query_matcher
    args:
      domain:
        - "provider:geosite:category-ads-all"

看来geosite里面确实有去广告规则，我直接用的passwall自带的geosite。
